### DISCUSSION TITLE
CommaFeed 3.0

### DISCUSSION THREAD

**Discussion by Athou** [Athou/commafeed#1015] [Announcements]
CommaFeed 3.0
-------------

Hey there,

I've been working for the past few weeks on CommaFeed 3.0. This version contains a complete rewrite of the UI.
I wanted to:
- have a more modern look while preserving the spirit of the previous UI
- replace the deprecated angularjs framework
- learn something new

The new UI is using React and Mantine and is written entirely in Typescript and I'm very happy with the result.

There are some features that are not yet implemented but it is perfectly usable on a day to day basis.
Here are the current missing features:
- [x] Expanded view for entries
- [x] Stars
- [x] Tags
- [x] Search
- [x] Sharing buttons
- [x] Right-to-left support

The code resides in the [develop](https://github.com/Athou/commafeed/tree/develop) branch. A prebuilt version is also available after every push [here](https://github.com/Athou/commafeed/actions).
There are no database changes and no configuration changes with this version so you can switch between the 3.0 version and the 2.6 version without any issues.
Note that if you build from sources, there is a slight change in the directory structure. The client and the backend are now separated in their own module. 
You still build the application the same way but the resulting jar is now located in `commafeed-server/target/` instead of `target/`

This is still a work in progress, let me know what you think :)

![light](https://user-images.githubusercontent.com/1256795/184886797-d5459aa8-974b-4ff8-81a6-19b5f6c14e52.png)
![dark](https://user-images.githubusercontent.com/1256795/184886828-1973f148-58a9-4c6d-9587-ee5e5d3cc2cb.png)

**Comment by ebraminio** [comment]
Fantastic! Looks just great. While ago I drafted something at https://ebraminio.github.io/reader/ feel free to check if is there anything worthy to pick up, maybe clicking on that day/night button at the top right or search button (the only half working buttons) gives something worth to pick up or not. Thanks!

**Reply by Athou** [reply]
I wondered about putting the theme switcher in the toolbar but I decided to put it in the dropdown as I rarely change the theme and I didn't feel like it was useful to see it at all times 😄 
![image](https://user-images.githubusercontent.com/1256795/184972801-263690e3-5085-46df-87c3-ecbcfd3c87f5.png)

**Reply by travelswagger** [reply]
It looks pretty, but it's very slow and almost unusable now. Is there an even more minimal view that can be deployed to speed up performance? 

I also miss the previous and next buttons that were part of the mobile web view. I can only swipe up and down the feed to scroll through the feed, which is more time consuming than a tap of previous/next feed story.

**Comment by testman42** [comment]
I like how it looks. :+1:

**Comment by CPUGPU** [comment]
Reserving judgement just yet.  Where are the stars on the line items?  Looks like a significant loss of information density.  Will there still be a "dense" one-line option for the right pane?  How narrow can the left pane go?  Looks too wide.  Thanks.

**Reply by Athou** [reply]
Thanks for the feedback!
Stars are not yet supported, as indicated in the original post. I agree with you the layout is a little less dense. I didn't plan to make a more compact design but I'll have a look it might not be a lot of work. Same for the width of the tree.

**Reply by Athou** [reply]
I added support for a compact layout, this is what it looks like right now:
![image](https://user-images.githubusercontent.com/1256795/191191658-6fa1d13c-594a-42fb-8606-fbdfb199beb9.png)

It's a little bit closer to what CommaFeed 2.x looks like right now :)

**Reply by CPUGPU** [reply]
That's looking better.  Thank you.  Simplicity, density, and keyboard control are why I switched to CommaFeed years-ago when google killed Reader.  I tried all the rest and only CommaFeed fit the need.  (Also why I have a reminder on my calendar to send financial support every year since 2013!)

**Comment by ebraminio** [comment]
As noted already RTL isn't on par, for example on this feed, https://fa.wikipedia.org/w/api.php?action=featuredfeed&feed=potd&feedformat=atom this is what CommaFeed.com looks like

<img width="1428" alt="image" src="https://user-images.githubusercontent.com/833473/186665066-5f16dff3-8883-4566-bfe1-0e9182bf58b3.png">

And this is what the upcoming version looks like,

<img width="1434" alt="image" src="https://user-images.githubusercontent.com/833473/186665165-d4e5f0ff-4ff8-4f56-a0f3-45e6b59828de.png">

**Reply by Athou** [reply]
Does this look better?
![image](https://user-images.githubusercontent.com/1256795/186674918-1e1224bb-d6df-45d1-9991-8d8a17086118.png)

**Reply by ebraminio** [reply]
Yes this is acceptable now, thanks :)

**Reply by Athou** [reply]
Ok I'll push something tomorrow. Maybe I can do even better by switching the whole layout to RTL if the selected language in the user profile is RTL with something like https://mantine.dev/guides/rtl/. What do you think?

**Reply by ebraminio** [reply]
That will be just fantastic, not that required, at least, I use English for feed reader UI, but it will be a great experience to have for you generally as it will be fun and interesting challenge.

**Comment by ebraminio** [comment]
Apparently when posts are already read, j/k n/p short keys won't work anymore.

**Reply by Athou** [reply]
j/k shortcuts should now work correctly, I still need to implement n/p

**Reply by Athou** [reply]
n/p shortcuts should now work

**Comment by ebraminio** [comment]
Build from source instructions should be tweaked a bit I guess https://github.com/Athou/commafeed/tree/develop#the-short-version-build-from-sources as there is no `config.yml.example` there apparently

**Comment by Athou** [comment]
I finished my TODO list I think I implemented all the features of CommaFeed 2.6 in CommaFeed 3.0.
I released RC1 here https://github.com/Athou/commafeed/releases/tag/3.0.0-rc1 🎉

**Comment by junpet** [comment]
I'm a big fan of Commafeed, thanks your work. I use it every day, mostly on desktop, but I check on it on my smartphone as well. 

Since I use mostly on desktop, I'd say the compact mode still too big and waste of space for me (overall everything is too big, the fonts, except the counters). On the other hand on mobile it looks really neat.

What I miss is the starring. With the new frontend you need 4 clicks. Click on the article, make it unread, starring and click on the title again to close the article. For a long article its not pleasant. On desktop there's plenty space for a star icon, but on mobile it isn't, because everything got bigger. To be honest I don't know what would be the best option. Maybe have starring option on desktop and not on mobile? But starring would be much more cumbersome on mobile. I think best bet would be an option to toggle on/off starring.

I only used the dark theme, but something is off. After a little time I start seeing stripes.

This feedback is born after using the new frontend only for some minutes. I'm going to use it and check back later. I'm a little sad I only found about this just now, after releasing rc1.

Anyway, thanks your hard work! Nice job!

**Reply by Athou** [reply]
Thanks for the feedback, I'll see what I can do!

**Reply by junpet** [reply]
Thanks for your effort! 

One more thing. Could you make disappear read items after clicking `Mark all as read` button?

**Reply by junpet** [reply]
Another one. Let us disable shortcuts. I marked everything read, fortunately I had backup.

**Comment by flisk** [comment]
Hi! I've been a long time user of CommaFeed, and I'm very happy to see it's still actively maintained. Thanks for your work, @Athou :)

Some thoughts on the new UI:
- I'm a macOS user, and my OS automatically adjusts its color scheme based on the time of day. It would be nice if there was an option to have CommaFeed's color scheme follow my system as well (using a `prefers-color-scheme` media query) instead of having to change this manually.
- I read a lot of the feeds I'm subscribed to on the site that publishes them. My workflow for this in the old UI was fairly simple: click on the item in CommaFeed, click on the headline. In the new UI, I have to scroll all the way past the inline post after opening it to get to the "Open link" button, which is a little cumbersome. Making titles into links to the post again would be the simple solution of course, but I wonder if the controls at the bottom of expanded posts couldn't be placed in a more consistently accessible position.
- I also concur with @junpet about the dark theme – I don't find it as easily "parseable" as the light variant for some reason.

I'm excited to see where you'll take this project going forward!

**Comment by moneytoo** [comment]
Hi. I'm trying out CommaFeed and I'm really liking it! When I first ran the 3.0.0 version, I noticed quite low contrast of some elements. The contrast of the orange text on white background is 2.57:1, for orange on light orange it is 2.36:1. Lighthouse test confirms the same accessibility issue. I understand that this is the accent color for this app but maybe it needs to be a little darker.

**Comment by canoine** [comment]
Nice work, @Athou. :1st_place_medal: 

After some minor adjustments, all seems to work well on my venerable Core2Quad server, even if the compilation time has significantly increased.
I just had to change the target directory variable into my installation script, and... the java.tmpdir environment variable for maven, because my /tmp FS is mounted noexec, and the jansi library doesn't like that at all...

I just hope the runtime memory consumption won't explode.

The new interface is nice, even if the "compact" mode is unfortunately not as compact as the old one. As I use the "Dark reader" firefox module, the dark mode is useless for me, but I appreciate the addition.

Thank you. :)

**Comment by CPUGPU** [comment]
Long time user of Commafeed, I've been using 3.x since it was rolled-out to the commafeed.com instance and now have some comments:

1. It's slow.  If there are about 100 unread articles, the "All" top level never loads.  The throbber just keeps going.  I've waited up to an hour, I think.  I have 112 subscriptions total.  If I select a subcategory, it's much faster, but frequently still takes many minutes before I get an article list.
2. Display density preferences are not saved.  Every time I log in, I have to choose "compact" again.  This may be because I always use incognito mode browser.
3. Stars are not visible in-line in the article list.  They were in 2.x.

Happy to open an issue for any of these if-needed.  
Thank you for your continued effort!  Still the best feedreader, IMHO. :+1:

**Reply by Athou** [reply]
> 1. It's slow.  If there are about 100 unread articles, the "All" top level never loads.  The throbber just keeps going.  I've waited up to an hour, I think.  I have 112 subscriptions total.  If I select a subcategory, it's much faster, but frequently still takes many minutes before I get an article list.

I'm not sure why it is slower than the previous version. Nothing of importance changed in the backend since 2.6.0.
However, the server is very old and is using very slow hard drives. It may be time to upgrade to a server with SSD drives.
commafeed.com is not generating any revenue though, it has always been a side project and offered as a courtesy, so I'm a bit reluctant in increasing its monthly cost, since the server costs come from my own pocket.



> 2. Display density preferences are not saved.  Every time I log in, I have to choose "compact" again.  This may be because I always use incognito mode browser.

Yes, the setting is stored in the local storage of the browser. It won't be saved if you're browsing in icognito mode.


> 3. Stars are not visible in-line in the article list.  They were in 2.x.

Feel free to open an issue for this one :)

**Reply by CPUGPU** [reply]
Don't upgrade the server for me!  You are under no obligation to continue hosting it, in my opinion.  I appreciate the courtesy use of it, and have contributed every year since 2013 to show my appreciation.  I'll continue to support the project even if the server goes away.  (On my task list to host my own instance, but just lazy!)

I'll open an issue for the stars.   Thank you!

**Reply by Athou** [reply]
I managed to find a server with SSD drives at the same cost as the current server.
I migrated users and feeds but I could not migrate entries because the old drives were so slow that it would have taken weeks.

Can you give it a try https://new.commafeed.com/ ? It should be **way** faster than the current server :)

I'll monitor things for a while, if everything is smooth I'll consider switching commafeed.com to this instance.

**Comment by ampersand0415** [comment]
Sadly I have to agree with CPUGPU. I'm a long time user of CommaFeed and really appreciate your work but the latest update has essentially broken the website for me. It takes minutes to get past the Loading Commafeed (always sticks at 75%) and feeds with a large amount of posts don't even load at all.

**Comment by CPUGPU** [comment]
Following up on this discussion for future readers, I responded to Athou about the excellent new server here:
https://github.com/Athou/commafeed/issues/1125#issuecomment-1660123127

Summary:  The difference is amazing; the new server is near-instant response.  Athou will formally replace the old server after verify the new server is running well for a few days.  Users and settings will migrate, but old articles will not due to constraints of the old server.

Many thanks to Athou for taking the time to find and set up the new server!

**Comment by ampersand0415** [comment]
I can also confirm the functionality and speed is infinitely better than on the previous server. Thanks, Althou!
